src/ModelContextProtocol.Core/Server/McpServerFilters.cs

```diff
@@ -0,0 +1,161 @@
+using ModelContextProtocol.Protocol;
+
+namespace ModelContextProtocol.Server;
+
+/// <summary>
+/// Provides filter collections for MCP server handlers.
+/// </summary>
+/// <remarks>
+/// This class contains collections of filters that can be applied to various MCP server handlers.
+/// This allows for middleware-style composition where filters can perform actions before and after the inner handler.
+/// </remarks>
+public sealed class McpServerFilters
+{
+    /// <summary>
+    /// Gets the filters for the list tools handler pipeline.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// These filters wrap handlers that return a list of available tools when requested by a client.
+    /// The filters can modify, log, or perform additional operations on requests and responses for
+    /// <see cref="RequestMethods.ToolsList"/> requests. It supports pagination through the cursor mechanism,
+    /// where the client can make repeated calls with the cursor returned by the previous call to retrieve more tools.
+    /// </para>
+    /// <para>
+    /// These filters work alongside any tools defined in the <see cref="McpServerTool"/> collection.
+    /// Tools from both sources will be combined when returning results to clients.
+    /// </para>
+    /// </remarks>
+    public List<Func<Func<RequestContext<ListToolsRequestParams>, CancellationToken, ValueTask<ListToolsResult>>, Func<RequestContext<ListToolsRequestParams>, CancellationToken, ValueTask<ListToolsResult>>>> ListToolsFilters { get; } = new();
```

Comment:
The amount of generics nesting in this is going to trigger some codebases that have CA rules to prevent it. Would be better to have declarative delegate types that are well named.